### DISCUSSION TITLE
use with guard / exceptions filters?

### DISCUSSION THREAD

**Discussion by tal-rofe** [jmcdo29/nest-commander#453] [Q&A]
I wonder whether is it possible.
I want to set a global exception filter.

this is my `main.ts` file:
```ts
import { CommandFactory } from 'nest-commander';

import { AppModule } from './app.module';

async function bootstrap() {
	await CommandFactory.run(AppModule);
}

bootstrap();
```

this is the `app.module.ts` file
```ts
import { Module } from '@nestjs/common';
import { APP_FILTER, APP_GUARD } from '@nestjs/core';

import { AuthCommand } from './commands/auth/auth.command';
import { UnknownExceptionsFilter } from './filters/unknown.filter';
import { ConnnectionGuard } from './guards/connection.guard';
import { ConfigModule } from './modules/config/config.module';

@Module({
	imports: [ConfigModule],
	providers: [
		{
			provide: APP_FILTER,
			useClass: UnknownExceptionsFilter,
		},
		{
			provide: APP_GUARD,
			useClass: ConnnectionGuard,
		},
		AuthCommand,
	],
})
export class AppModule {}

```

The exception filter
```ts
import { Catch, ExceptionFilter } from '@nestjs/common';

@Catch()
export class UnknownExceptionsFilter implements ExceptionFilter {
	public catch() {
		console.log(5);
	}
}
```

Then I try to fail the program and throw an exception intentionally
```ts
import { Command, CommandRunner } from 'nest-commander';

@Command({ name: 'auth', description: 'Authenticate t' })
export class AuthCommand implements CommandRunner {
	public async run() {
		await Promise.resolve();

		throw new Error();
	}
}
```

But the exception handler does not do anything. Why?

**Comment by jmcdo29** [comment]
Enhancers (guards, pipes, interceptors, and filters) currently only work in contexts that have an idea of a "request". This is because it requires the `NestApplication` rather than a `NestApplicationContext`. To have this though, there is a requirement to include an HTTP adapter which is overkill for the context of CLI requests. For error handling, you can just add a `try/catch` around the command

**Reply by tal-rofe** [reply]
@jmcdo29 

It is something to be supported in this package in the future?
That;d make the package x10 better..

**Reply by jmcdo29** [reply]
I'll see if I can think of a way to include something like this. It'll take some time to work out how it can all work together, and I'll have to re-build some of Nest's internals inside this package to handle the use of guards and filters. I will not be adding pipe or interceptor support at the moment though.

With regards to the `ExecutionContext` and `ArgumentHost` objects, what do you see as necessary values to have in the guards and filters?

**Reply by tal-rofe** [reply]
@jmcdo29 
Also, injecting services for example does not work.
I would make PRs if I had the time.

I can't tell how this package is related with NestJS if most of the NestJS's features are disabled.
I think the most important ones are:
- Injecting services
- Having **Guards** and **Filters**

**Reply by jmcdo29** [reply]
Can you provide a reproduction that shows that you can't inject services? I have [examples in my test suite](https://github.com/jmcdo29/nest-commander/blob/main/integration/this-command/src/this-command.command.ts) that shows injection works just fine.

**Reply by smolinari** [reply]
> @jmcdo29
> 
> It is something to be supported in this package in the future? That;d make the package x10 better..

For my curiosity, can you also expand on how somehow shoehorning in a request context would make the package x10 better? What would it allow for, which you can't do now? What is it you are trying to accomplish, above and beyond catching errors (which Jay already explained what you can do)?

Scott

**Reply by tal-rofe** [reply]
@smolinari 

Regarding the Filters, I want to have a global one. I also want to catch specific custom exceptions.
Regarding the Guards, I want to have some commands which require internet connection. For that case, I'd set a guard which checks whether the client has connection.

**Reply by smolinari** [reply]
@tal-rofe 
>  I want to have some commands which require internet connection. 

Somehow this makes no sense to me. If your command is needing to connect to another API, then your command is the client. Right?

Scott

**Reply by tal-rofe** [reply]
@smolinari 

What's wrong here? Imagine dummy CLI renders names of students off some college API.

**Reply by jmcdo29** [reply]
I don't think I see the need for a guard, just try to make the request and if no internet connection then print that fact out, yeah? 

```ts
@Command({ name: 'some-command' })
export class SomeCommand implements CommandRunner {
  async run() {
    try {
      const res = await fetch(someUrl)
      const data = await res.json()
      console.log(doSomethingWithData(data))
    } catch (err) {
      console.error('Something went wrong');
      if (err.statusText === 'Connection Refused') {
        console.error("It seems you aren't connected to the internet. Please re-connect and try again")
      } else {
        console.error(err)
      }
    }
  }
}
```

**Reply by tal-rofe** [reply]
But why? Isn't it exactly the use for guard? I could build a guard for guarding commands with internet connectivity and apply it globally with Nestjs core features.

Why doesn't this implementation satisfy?

**Reply by jmcdo29** [reply]
Like I said before, to make enhancers work, Nest needs an HTTP context (or some other external request like websocket or rpc). In the [Standalone Context](https://docs.nestjs.com/standalone-applications) that nest-commander makes use of, enhancers don't work. This is how Nest is designed. 

As I said before, I can work on making guards and filters usable, but I'll need to re-build some of Nest's code and I need to determine what will be expected in the `ExecutionContext` and `ArgumentHost` for the guards and filters.

**Reply by tal-rofe** [reply]
@jmcdo29 

Regarding the injection of services that does not work for me,
when I try to inject a service into the command handler (`@Command`) nothing happens.

My `main.ts` file
```ts
import { CommandFactory } from 'nest-commander';

import { AppModule } from './app.module';

async function bootstrap() {
	await CommandFactory.run(AppModule);
}

bootstrap();
```

My `app.module.ts` file:
```ts
import { Module } from '@nestjs/common';

import { AuthCommand } from './commands/auth/Auth';
import { ConfigModule } from './modules/config/config.module';
import { ConnectionModule } from './modules/connection/connection.module';

@Module({
	imports: [ConfigModule, ConnectionModule],
	providers: [AuthCommand],
})
export class AppModule {}
```

My `config.module.ts` file:
```ts
import { Global, Module } from '@nestjs/common';
import { ConfigService } from './config.service';

@Global()
@Module({ providers: [ConfigService] })
export class ConfigModule {}
```

My `config.service.ts` file:
```ts
import { Injectable } from '@nestjs/common';
import { IConfig } from './config.interface';

@Injectable()
export class ConfigService {
	private _config: IConfig = {
		API_URL: 'blabla',
	};

	public getValue(key: keyof IConfig) {
		return this._config[key];
	}
}
```

Then, I try to inject the `ConfigService` within my command handler:
```ts
import { Command, CommandRunner } from 'nest-commander';

import { ConfigService } from '@/modules/config/config.service';
@Command({ name: 'auth', description: 'Authenticate dssfds CLI with an sdfdsf account' })
export class AuthCommand implements CommandRunner {
	constructor(private readonly configService: ConfigService) {}

	public async run() {
		const hasConnection = await this.configService.getValue('API_URL');

		console.log('blabla');
	}
}
```

When I run: `node dist/main.js auth` nothing happens. When I remove the injection:
```ts
```ts
import { Command, CommandRunner } from 'nest-commander';

import { ConfigService } from '@/modules/config/config.service';
@Command({ name: 'auth', description: 'Authenticate fddf CLI with an dfssdf account' })
export class AuthCommand implements CommandRunner {

	public async run() {

		console.log('blabla');
	}
}
```

Then if I run the same command again, I do see the message logged to the console.

**Reply by jmcdo29** [reply]
Your `ConfigModule` does not have the `ConfigService` exported. Nest is logging an error, but `nest-commander` sets `logger: false` to keep the command logs clean. You can set a logger yourself using the `logger` option just like you normally would with `NestFactory.create` and see that there's an injection error about not being able to find the `ConfigService` provider. 

[docs](https://nest-commander.jaymcdoniel.dev/docs/features/factory/#logging)

**Reply by smolinari** [reply]
> But why? Isn't it exactly the use for guard? I could build a guard for guarding commands with internet connectivity and apply it globally with Nestjs core features.
> 
> Why doesn't this implementation satisfy?

@tal-rofe - I'm still confused. If you are creating a command, you'd be calling an API, not expecting requests from it, only responses, right?

Scott

**Reply by tal-rofe** [reply]
@jmcdo29  Thanks, issue was resolved.

@smolinari 

I think that a command handler is similar to controller handler. That's why I thought of apply guard and filters
